man/spectrum.cfg.5
.\"
.\"     Title: spectrum.cfg
.\"    Author: Mathias Ertl <mati@fsinf.at>
.\"  Language: English
.\"      Date: 2010-03-31
.\"
.TH spectrum.cfg 5  "March 31, 2010" "Version 0.1\-git" "Spectrum Manual"

.SH NAME
spectrumc.cfg - spectrum configuration file
.SH FULL PATH
Depends on local setup, but usually in /etc/spectrum.
.SH DESCRIPTION
This man-page describes the fileformat for spectrum configuration files. With
\fBspectrum\fR, each protocol configuration (e.g. for icq, msn, ...) runs as its
own instance and hence requires its own configuration file.

The fileformat is structured into four sections: \fIservice\fR, \fIlogging\fR,
\fIdatabase\fR and \fIpurple\fR. The beginning is marked by the section enclosed
in brackets, here is an example:

.RS
[service] 
.br
# lines starting with a '#' are considered a comment.
.br
enable=1
.br
protocol=xmpp
.br
\&...
.br
[logging]
.RE
.SH FILENAME
The filename itself can be used to configure a few configuration details. This
is intended to enable running all \fBspectrum\fR instances with identical
configuration files that only differ in their filename. You can use this feature
to have only one file and symlink (see ln(1)) it for each instance. 

The syntax of the filename is:
.RS 4
.sp
\fBjid[:protocol[:port]].cfg\fR or
.br
\fBprotocol:port.cfg\fR
.sp
.RE
So a full filename would be \fBxmpp.example.com:xmpp:1000.cfg\fR or, if only two
fields are given, \fBxmpp:1000.cfg\fR. You can use these settings by using the
strings "$filename:jid" (if given), "$filename:protocol" and "$filename:port"
(without quotes) in the configuration file.
.SH COMMON OPTIONS
Most options are common to all protocol configurations. Settings can only be
omitted if there is a default value or if it is explicitly marked as optional.

.SS SECTION service
\fBenable\fR=\fIbool\fR
.RS
To disable the spectrum instance, set \fIbool\fR to \fI0\fR (default: enable=1).
.RE

\fBprotocol\fR=\fIprotocol\fR
.RS
Have this spectrum instance use \fIprotocol\fR. Valid values are \fIaim\fR,
\fIfacebook\fR, \fIgg\fR, \fIicq\fR, \fIirc\fR, \fImsn\fR, \fImsn_pecan\fR,
\fImyspace\fR, \fIqq\fR, \fIsimple\fR, \fItwitter\fR, \fIxmpp\fR or \fIyahoo\fR.
.RE

\fBserver\fR=\fIaddress\fR
.RS
Your XMPP-server is available at \fIaddress\fR, the setting can be any IPv4- or
IPv6-address or a fully qualified domain name. Most of the time \fIaddress\fR will
just be \fI127.0.0.1\fR.

\fBNote:\fR IPv6 support is currently untested.
.RE

\fBport\fR=\fIport\fR
.RS
Use \fIport\fR as the component port.
.RE

\fBjid\fR=\fIjid\fR
.RS
Use \fIjid\fR as the component JID.
.RE

\fBuse_proxy\fR=\fIbool\fR
.RS
If \fIbool\fR is 1, the http_proxy env var will be used as the proxy server
(default: 0).
Example: 
.RS
http_proxy="http://user:passwd@your.proxy.server:port/"
.RE
.RE

\fBconfig_interface\fR=\fIpath\fR
.RS
Spectrum opens a local socket at \fIpath\fR where it replies to some 
XMPP-packets. This setting is used by the spectrumctl-action \fIstats\fR
and the munin plugin. The default is /var/run/spectrum/$jid.sock
.RE

\fBfiletransfer_bind_address\fR=\fIIP\fR:\fIport\fR
.RS
The filetransfer proxy will bind to \fIIP\fR:\fIport\fR. \fIIP\fR has
to be a public IP. Example: 192.0.2.1:12345.
Only used when filetransfer is enabled (see \fItransport_features\fR and
\fIvip_features\fR settings).
.RE

\fBfiletransfer_public_address\fR=\fIIP\fR:\fIport\fR
.RS
The \fIIP\fR and \fIport\fR that will be sent as a stream host upon a 
filetransfer request. Example: 192.0.2.1:12345.
Only used when filetransfer is enabled (see \fItransport_features\fR and
\fIvip_features\fR settings).
.RE

\fBfiletransfer_cache=\fR=\fIdirectory\fR
.RS
Downloaded files will be saved at \fIdirectory\fR.
Only used when filetransfer is enabled (see \fItransport_features\fR and
\fIvip_features\fR settings).
.RE

\fBfiletransfer_web\fR=\fIURL\fR
.RS
Downloaded files will be available at \fIURL\fR. Note that it is your job to
configure your webserver accordingly.
.RE

\fBadmins\fR=\fIuser@example.com,...\fR
.RS
Optional. Comma-seperated list of transport administrators who have access to
admin adhoc commands.
.RE

\fBname\fR=\fIname\fR
.RS
The transport will show up in Service Discovery as \fIname\fR.
.RE

\fBlanguage\fR=\fIlang\fR
.RS
Use \fIlang\fR as the default language for spetrum (default: en).
.RE

\fBtransport_features\fR=\fIfeature\fR;\fI...\fR
.RS
List of features available on the transport, seperated by semicolons (';'),
\fIfeature\fR must be a combination of \fIavatars\fR, \fIchatstates\fR and
\fIfiletransfer\fR. By default, spectrum offers all three features.
.RE

\fBvip_mode\fR=\fImode\fR
.RS
If \fImode\fR is \fI1\fR, users will be devided into two groups based
on a database table. See also: \fIonly_for_vip\fR, \fIallowed_servers\fR and
\fIvip_features\fR (default: 0).
.RE

\fBonly_for_vip\fR=\fIbool\fR
.RS
If both \fIvip_mode\fR and \fIonly_for_vip\fR are set to \fI1\fR, the transport
will only be available to VIP users (default: 0).
.RE

\fBallowed_servers\fR=\fIserver\fR;\fI...\fR
.RS
Optional. On a VIP only transport, accounts on any of the given \fIservers\fR
are able to use the transport even if they are not VIP users.
.br
This feature is useful if you want to enable transport only for users from your
server, but also want to give access to VIP users from other servers (for example
from GTalk).
.RE

\fBvip_features\fR=\fIfeature\fR;\fI...\fR
.RS
List of features available only for VIP users. Ignored if
\fIvip_mode\fR=\fI0\fR. List of features is the same as in
\fItransport_feauters\fR.
.RE

\fBpid_file\fR=\fIpath\fR
.RS
Spectrum will put its PID file at \fIpath\fR (default: /var/run/spectrum/$jid). 
.RE

.SS SECTION logging
\fBlog_file\fR=\fIpath\fR
.RS
Create log-file at \fIpath\fR.
.RE

\fBlog_areas\fR=\fIpurple\fR;\fIxml\fR
.RS
A combination of \fIpurple\fR and \fIxml\fR.
.RE

.SS SECTION database
\fBtype\fR=\fItype\fR
.RS
Either use \fIsqlite\fR or \fImysql\fR as database backend.
.RE

\fBhost\fR=\fIhostname\fR
.RS
Connect to \fIhostname\fR when using \fImysql\fR.
.RE

\fBuser\fR=\fIname\fR
.RS
Use \fIname\fR as username when for connecting to \fImysql\fR.
.RE

\fBpassword\fR=\fIpwd\fR
.RS
Use \fIpwd\fR as password when for connecting to \fImysql\fR.
.RE

\fBdatabase\fR=\fIdb\fR
.RS
When using \fImysql\fR, use \fIdb\fR as database name, when using \fIsqlite\fR,
use \fIdb\fR to set the path to the database file.
.RE

\fBprefix\fR=\fIprefix\fR
.RS
Optional. Use \fIprefix\fR as database prefix to connect to a \fImysql\fR
database.
.RE

.SS SECTION purple
\fBuserdir\fR=\fIdirectory\fR
.RS
Store dynamic user data in \fIdirectory\fR.
.RE

.SH XMPP specific options
.SS SECTION service
\fBrequire_tls\fR=\fIbool\fR
.RS
Require TLS to connect to the legacy network. Set this to \fI0\fR to enable
connectivity to the Facebook Jabber server (default: \fI1\fR).
.RE

.SH AUTHOR
Copyright \(co 2009\-2010 by Spectrum engineers:
.sp
.\" template start
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Jan Kaluza <hanzz@soc.pidgin\&.im>
.RE
.\" template end, and once again template start
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Mathias Ertl <mati@fsinf\&.at>
.RE
.\" template end ;)
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Paul Aurich <paul@darkrain42\&.org>
.RE
.\" again template end

License GPLv3+: GNU GPL version 3 or later.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.sp
See http://gnu.org/licenses/gpl.html for more information.
.SH SEE ALSO
\fBspectrum\fP(1), \fBspectrumctl\fP(8)

.SH BUGS
Please submit bugs to our issue tracker at github:
http://github.com/hanzz/spectrum/issues


ChangeLog
version 0.3 (2010-??-??):
	General
	* Try to reconnect to Jabber server for all stream errors.
	* Disconnect users from legacy network when Jabber server gets down.

	IRC:
	* Fixed bug when user stays connected to IRC after he logged out from Jabber.
	* Fixed sending first private message from XMPP to IRC.

version 0.2 (2010-04-08):
	General:
	* Don't answer error IQ stanza on disco#info request.
	* Added Windows compilation support.
	* Send unavailable presences when shutting transport down.
	* Working filetransfer from XMPP to legacy network when legacy network user
	  can't receive file. File is then stored on server and the link to file is
	  sent. Web address and file storage directory can be set in config file.
	* Twitter protocol support (using http://code.google.com/p/prpltwtr/)
	* Added $filename:jid, $filename:protocol and $filename:port config variables
	  which are set according to config filename. It has to be named as
	  jid:protocol:port.cfg (for example icq.localhost:icq:5432.cfg).
	* Fixed crash when legacy network user cancels filetransfer.
	* Fixed crash when XMPP user cancels Adhoc command execution.
	* Fixed crash when two XMPP users try to use one legacy network account.
	* Don't send duplicate presences.
	* Don't forward typing notification without an existing conversation.
	* Fixed transport unregister process. Buddies are now removed also if user
	  doesn't support Roster Item Exchange.
	* Forward VCard avatars from XMPP to legacy network, with the caveat that
	  this only works if the VCard avatar adheres to the file type, size,
	  and dimension constraints of the legacy network (no scaling is currently
	  done).
	* Added filetransfer_bind_address and filetransfer_public_address config
	  variables to be able to set IP and port for filetransfer proxy.  The
	  public address is used if Spectrum is behind a NAT.
	* Spectrum can now open a local socket that allows (fast) statistics
	  gathering. This is already used by spectrumctl. The path to the
	  local socket is configured with the config_interface option.
	* New man pages for spectrum.cfg and spectrumctl.

	spectrumctl:
	* The script now does rigorous checking of ownership and permissions
	  on all the important files and directories.
	* New action "stats" which gathers statistics via the config_interface
	  (see above).
	* Two new command-line options for start: 
	  * --no-daemon: Do not start spectrum as a daemon
	  * --debug: Sets some environment variables, a coredump will be made
	    to $HOME in case of a segfault.
	* New command-line option for list: 
	  * --machine-readable: Output data as comma-seperated values
	* env-variable SPECTRUM_PATH can now configure how spectrumctl is
	  called. If the variable is not set, it assumes that spectrum is in its
	  PATH.
	* env-variable SPECTRUM_USER can now configure the name of the
	  spectrum-user. This is still overridden by --su.

	XMPP:
	* MUC support for XMPP

	IRC:
	* IRC Nickserv password can be set via Adhoc command.
	* Fixed Private Messages.
	* Using IRC Helper (http://plugins.guifications.org/trac/wiki/irchelper) for
	  identyfing user. Used only if IRC Helper is installed.

	Munin plugin
	* Now use config_interface instead of XMPP-connection. This changes
	  the configuraton significantly, please see documentation of the
	  plugin for details.

version 0.1 (2010-02-05):
	* initial version


spectrumctl/spectrum/spectrum.py
# -*- coding: utf-8 -*-
#
# This file is part of spectrumctl. See spectrumctl.py for a description.
#
# Copyright (C) 2009, 2010 Mathias Ertl
#
# Spectrumctl is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <http://www.gnu.org/licenses/>.

import os, grp, pwd, stat, time, signal, subprocess
try:
	from spectrum import spectrumconfigparser
except ImportError:
	import spectrumconfigparser

class ExistsError( RuntimeError ):
	"""Thrown when a file or directory does not exist."""
	pass

class spectrum:
	def __init__( self, options, config_path ):
		self.options = options
		self.config_path = os.path.normpath( config_path )
		
		self.config = spectrumconfigparser.SpectrumConfigParser()
		self.config.read( config_path )
		self.pid_file = self.config.get( 'service', 'pid_file' )

	def get_jid( self ):
		"""
		Get the jid of this service
		"""
		return self.config.get( 'service', 'jid' )

	def enabled( self ):
		return self.config.getboolean( 'service', 'enable' )

	def get_pid( self ):
		"""
		Get the pid of the service, returns -1 if pid does not exist or
		is unparsable.
		"""
		try:
			return int( open( self.pid_file ).readlines()[0].strip() )
		except:
			return -1

	def check_exists( self, file, typ=file ):
		if not os.path.exists( file ):
			raise ExistsError( file, 'Does not exist' )

		if typ == 'file' and not os.path.isfile( file ):
			raise RuntimeError( file, 'Not a file' )
		if typ == 'dir' and not os.path.isdir( file ):
			raise RuntimeError( file, 'Not a directory' )

		return True

	def check_ownership( self, node, uid=None, gid=None ):
		"""
		Check the ownership and group ownership of the given filesystem
		node. If uid or gid is None, the uid (and its respective primary
		gid) is used. If you don't want to check uid or gid, just pass
		-1.
		"""

		if uid == -1 and gid == -1:
			print( "Warning: uid and gid are -1. That doesn't make sense." )

		if not os.path.exists( node ):
			raise RuntimeError( node, "Does not exist" )

		if uid == None:
			uid = self.get_uid()
		if gid == None:
			gid = self.get_gid()

		stat = os.stat( node )
		if uid != -1 and stat.st_uid != uid:
			name = pwd.getpwuid( uid ).pw_name
			raise RuntimeError( node, 'Unsafe ownership (fix with "chown %s %s")'%(name, node) )
		if gid != -1 and stat.st_gid != gid:
			name = grp.getgrgid( gid ).gr_name
			raise RuntimeError( node, 'Unsafe ownership (fix with "chgrp %s %s")'%(name, node) )
	
	def perm_translate( self, perm ):
		"""
		Transforms one of the modes defined in the stats module into a
		string tuple ('x', 'y') where 'x' is one of 'u', 'g' or 'o' and
		'y' is either 'r', 'w' or 'x'. This method can be used transform
		a permission into string that can be used with chmod.
		"""
		if perm == stat.S_IRUSR:
			return ('u', 'r')
		elif perm == stat.S_IWUSR:
			return ('u', 'w')
		elif perm == stat.S_IXUSR:
			return ('u', 'x')
		elif perm == stat.S_IRGRP:
			return ('g', 'r')
		elif perm == stat.S_IWGRP:
			return ('g', 'w')
		elif perm == stat.S_IXGRP:
			return ('g', 'x')
		elif perm == stat.S_IROTH:
			return ('o', 'r')
		elif perm == stat.S_IWOTH:
			return ('o', 'w')
		elif perm == stat.S_IXOTH:
			return ('o', 'x')

	def chmod_translate( self, add, rem ):
		changes = []

		for who, what in add.iteritems():
			if what != []:
				changes.append( who + '+' + ''.join( what ) )
		for who, what in rem.iteritems():
			if what != []:
				changes.append( who + '-' + ''.join( what ) )

		return ','.join( changes )
	
	def check_permissions( self, file, permissions, wildcards=[] ):
		all = [ stat.S_IRUSR, stat.S_IWUSR, stat.S_IXUSR,
			stat.S_IRGRP, stat.S_IWGRP, stat.S_IXGRP,
			stat.S_IROTH, stat.S_IWOTH, stat.S_IXOTH ]
		mode = os.stat( file )[stat.ST_MODE]
		add = { 'u': [], 'g': [], 'o': [] }
		rem = { 'u': [], 'g': [], 'o': [] }

		for p in permissions:
			if (mode & p) == 0:
				who, what = self.perm_translate( p )
				add[who].append( what )

		for p in [ p for p in all if ( p not in permissions and p not in wildcards )]:
			if (mode & p) != 0:
				who, what = self.perm_translate( p )
				rem[who].append( what )

		mode_string = self.chmod_translate( add, rem )
		if mode_string:
			string = 'chmod ' + mode_string + ' ' + file
			raise RuntimeError( file, 'Incorrect permissions (fix with "%s")' %(string) )

	def get_uid( self ):
		# if we explicitly name something on the CLI, we use that:
		if self.options.su:
			return pwd.getpwnam( self.options.su ).pw_uid
			
		try:
			# if we have env-variable set, we use that:
			username = os.environ['SPECTRUM_USER']
		except KeyError:
			# otherwise we default to spectrum:
			username = 'spectrum'

		return pwd.getpwnam( username ).pw_uid

	def get_gid( self ):
		return pwd.getpwuid( self.get_uid() ).pw_gid

	def check_writable( self, node, uid=None ):
		if uid == None:
			user = pwd.getpwuid( self.get_uid() )
		else:
			user = pwd.getpwuid( uid )

		s = os.stat( node )
		groups = [g for g in grp.getgrall() if user.pw_name in g.gr_mem]
		groups.append( user.pw_gid )
		cmd = ''
		if s.st_uid == user.pw_uid:
			# user owns the file
			if s.st_mode & stat.S_IWUSR == 0:
				cmd = 'chmod u+w %s' %(node)
			if os.path.isdir( node ) and s.st_mode & stat.S_IXUSR == 0:
				cmd = 'chmod u+wx %s' %(node)
		elif s.st_gid in groups:
			if s.st_mode & stat.S_IWGRP == 0:
				cmd = 'chmod g+w %s' %(node)
			if os.path.isdir( node ) and s.st_mode & stat.S_IXGRP == 0:
				cmd = 'chmod g+wx %s' %(node)
		else: 
			if s.st_mode & stat.S_IWOTH == 0:
				cmd = 'chmod o+w %s' %(node)
			if os.path.isdir( node ) and s.st_mode & stat.S_IXOTH == 0:
				cmd = 'chmod o+wx %s' %(node)
		
		if cmd != '':
			raise RuntimeError( node, 'Not writable (fix with %s)' %(cmd) )
	
	def su_cmd( self, cmd ):
		user = pwd.getpwuid( self.get_uid() ).pw_name
		return [ 'su', user, '-s', '/bin/sh', '-c', ' '.join( cmd ) ]

	def create_dir( self, dir ):
		os.makedirs( dir )
		os.chmod( dir, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR | stat.S_IRGRP | stat.S_IXGRP )
		os.chown( dir, self.get_uid(), self.get_gid() )

	def check_environment( self ):
		# check if spectrum user exists:
		if os.name == 'posix':
			try:
				user = pwd.getpwuid( self.get_uid() )
			except KeyError:
				raise RuntimeError( self.get_uid(), 'UID does not exist' )
		else:
			# on windows, there is no real way to get the info that
			# we need
			return 0, "ok"
		
		# we must be root
		if os.getuid() != 0:
			raise RuntimeError( "User", "Must be root to start spectrum" )
		
		# check permissions for config-file:
		self.check_ownership( self.config_path, 0 )
		# we don't care about user permissions, group MUST be read-only
		self.check_permissions( self.config_path, [ stat.S_IRGRP ],
			[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IXUSR ] )

		# config_interface:
		config_interface = self.config.get( 'service', 'config_interface' )
		if config_interface:
			# on some old (pre 0.1) installations config_interface does not exist
			dir = os.path.dirname( config_interface )
			try: 
				self.check_exists( dir, 'dir' )
				self.check_writable( dir )
			except ExistsError:
				self.create_dir( dir )

		# filetransfer cache
		filetransfer_cache = self.config.get( 'service', 'filetransfer_cache' )
		try:
			self.check_exists( filetransfer_cache, 'dir' )
			self.check_ownership( filetransfer_cache )
			self.check_permissions( filetransfer_cache, # rwxr-x---
				[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IXUSR, 
				  stat.S_IRGRP, stat.S_IXGRP ] )
		except ExistsError:
			self.create_dir( filetransfer_cache )

		# pid_file:
		pid_file = self.config.get( 'service', 'pid_file' )
		pid_dir = os.path.dirname( pid_file )
		try:
			self.check_exists( pid_dir )
			self.check_writable( pid_dir )
		except ExistsError:
			self.create_dir( pid_dir )

		# log file
		log_file = self.config.get( 'logging', 'log_file' )
		try:
			# does not exist upon first startup:
			self.check_exists( log_file )
			self.check_ownership( log_file, gid=-1 )
			self.check_permissions( log_file, # rw-r-----
				[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IRGRP ] )
		except ExistsError:
			log_dir = os.path.dirname( log_file )
			try:
				self.check_exists( log_dir )
			except ExistsError:
				self.create_dir( log_dir )

		# sqlite database
		if self.config.get( 'database', 'type' ) == 'sqlite':
			db_file = self.config.get( 'database', 'database' )
			try:
				# might not exist upon first startup:
				self.check_exists( db_file )
				self.check_ownership( db_file )
				self.check_permissions( db_file, # rw-r-----
					[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IRGRP ] )
			except ExistsError:
				db_dir = os.path.dirname( db_file )
				try:
					self.check_exists( db_dir )
					self.check_ownership( db_dir )
					self.check_permissions( db_dir, # rwxr-x---
						[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IXUSR, 
						  stat.S_IRGRP, stat.S_IXGRP ] )
				except ExistsError:
					self.create_dir( os.path.dirname( db_file ) )
		
		# purple userdir
		userdir = self.config.get( 'purple', 'userdir' )
		try:
			self.check_exists( userdir, 'dir' )
			self.check_ownership( userdir )
			self.check_permissions( userdir, # rwxr-x---
				[ stat.S_IRUSR, stat.S_IWUSR, stat.S_IXUSR, 
				  stat.S_IRGRP, stat.S_IXGRP ] )
		except ExistsError:
			self.create_dir( userdir )

		return 0, 'ok'

	def list( self ):
		pid = self.get_pid()
		if pid == -1:
			pid = '-'
		proto = self.config.get( 'service', 'protocol' )
		host = self.get_jid()
		status = self.status()[1]
		return (pid, proto, host, status)

	def status( self, pid=None ):
		"""
		Determines if the instance is running.

		This method uses the /proc filesystem if it exists and tries
		to send signal 0 if not.

		@return: (int, string) where int is the exit-code and string is a status message.
		@see:	http://refspecs.freestandards.org/LSB_3.1.0/LSB-Core-generic/LSB-Core-generic/iniscrptact.html
		"""
		if not os.path.exists( self.pid_file ):
			return 3, "not running."

		if pid == None:
			pid = self.get_pid()
			if pid <= 0:
				return 4, "could not parse pid file."
		
		if os.path.exists( '/proc' ):
			if os.path.exists( '/proc/%i' %(pid) ):
				return 0, "running."
			else:
				return 1, "not running but pid file exists."
		else: # no proc!
			try:
				os.kill( pid, 0 )
				return 0, "running."
			except OSError, err:
			# except OSError as err: #python3
				if err.errno == errno.ESRCH:
					return 1, "not running but pid file exists."
				else:
					return 4, "unknown."
			except:
				return 4, "unknown."


	def start( self ):
		"""
		Starts the instance.

		@return: (int, string) where int is the exit-code and string is a status message.
		@see:	http://refspecs.freestandards.org/LSB_3.1.0/LSB-Core-generic/LSB-Core-generic/iniscrptact.html
		"""
		status = self.status()[0]
		if status == 0:
			return 0, "already running." # starting while we are already running is also success!
		elif status == 1:
			os.remove( self.pid_file )
		elif status != 3:
			return 1, "status unknown." # We cannot start if status != 3

		# do rigorous permission-checking:
		try:
			self.check_environment()
		except RuntimeError, e:
			return 1, "%s: %s" % e.args

		try:
			binary = os.environ['SPECTRUM_PATH']
		except KeyError:
			binary = 'spectrum'

		cmd = [ binary ]

		# get the absolute path of the config file, so we can change to
		# spectrums homedir
		path = os.path.abspath( self.config_path )
		home = pwd.getpwuid( self.get_uid() )[5]
		os.chdir( home )

		if self.options.no_daemon:
			cmd.append( '-n' )
		if self.options.debug:
			os.environ['MALLOC_PERTURB_'] = '254'
			os.environ['PURPLE_VERBOSE_DEBUG'] = '1'
			cmd[0] = 'ulimit -c unlimited; ' + cmd[0]
		cmd.append( path )
		cmd = self.su_cmd( cmd )
		retVal = subprocess.call( cmd )
		if retVal != 0:
			return 1, "could not start spectrum instance."

		return 0, "ok."

	def stop( self ):
		"""
		Stops the instance (sends SIGTERM).

		@return: (int, string) where int is the exit-code and string is a status message.
		@see:	http://refspecs.freestandards.org/LSB_3.1.0/LSB-Core-generic/LSB-Core-generic/iniscrptact.html
		"""
		status = self.status()[0]
		if status == 3:
			# stopping while not running is also success!
			return 0, "already stopped."
		elif status == 1:
			os.remove( self.pid_file )
			return 0, "pid file was still there"
		elif status != 0: 
			# we cannot stop if status != 0
			return 1, "status unknown." 

		pid = self.get_pid()
		try:
			os.kill( pid, signal.SIGTERM )
			time.sleep( 0.1 )
			
			for i in range(1, 10):
				status = self.status()[0]
				if status == 3 or status == 1:
					os.remove( self.pid_file )
					return 0, "ok."
				time.sleep( 1 )
				os.kill( pid, signal.SIGTERM )
		except:	
			return 1, "failed."

	def restart( self ):
		"""
		Restarts the instance (kills the process and starts it again).
		
		@return: (int, string) where int is the exit-code and string is a status message.
		@see:	http://refspecs.freestandards.org/LSB_3.1.0/LSB-Core-generic/LSB-Core-generic/iniscrptact.html
		"""
		if self.stop()[0] == 0:
			return self.start()
		else:
			return 1, "could not stop spectrum instance."

	def reload( self ): # send SIGHUP to process
		"""
		Reload instance (send SIGHUP) which causes spectrum to reopen
		log-files etc.
		
		@return: (int, string) where int is the exit-code and string is a status message.
		"""
		if self.status()[0] != 0:
			return 1, "not running."
		pid = self.get_pid()
		try:
			os.kill( pid, signal.SIGHUP )
			return 0, "ok."
		except:	
			return 1, "failed."
